pyro/ops/contract.py

```diff
@@ -95,7 +95,10 @@ class UnpackedLogRing(TensorRing):
     Ordinals are frozensets of ``CondIndepStackFrame``s.
     """
     def dims(self, term):
-        return [d for d in range(-term.dim(), 0) if term.size(d) > 1]
+        key = 'dims', id(term)
```

Comment:
previously we computed `dims` once and cached in `sum_dims`. Since we now call this more often, I've added memoization logic (as is already done in the other `TensorRing` implementation).